### PR TITLE
Parse Config File

### PR DESCRIPTION
**Pull request by mhgholamrezaei**
Changes:
- Add utility functions to the Utils module
- Added config file for DRAM
- Parse the config file to get the DRAM dimensions for functional simulation in pimDevice
- Parse the config file to get the DRAM all parameters for the performance model
- Get the config files path and names from the envirnment variables: 

According to the specification in #126 in old repo.

### CODE REVIEW

File: configs/PIMeval_BitSimdV.cfg
Comment (fasiddique) [inline configs/PIMeval_BitSimdV.cfg:7]: Let's remove this as it will be hardware dependent.

**Review by fasiddique — COMMENTED** [review]
@mhgholamrezaei thank you for doing this! Could you please add a test example so that I can test it?

File: configs/PIMeval_BitSimdV.cfg
Comment (deyuan) [reply configs/PIMeval_BitSimdV.cfg:7]: This is upper limit. The actual number of threads is also determined by hardware concurrency. 
Could we support comments in this config file, such as "#..."?

File: libpimeval/src/libpimeval.h
Comment (deyuan) [inline libpimeval/src/libpimeval.h:33]: We can use PIM_DEVICE_NONE for unknown cases

File: libpimeval/src/pimSim.cpp
Comment (deyuan) [inline libpimeval/src/pimSim.cpp:79]: We need either uninit or handle the case of calling init twice and avoid MLK

File: libpimeval/src/pimDevice.cpp
Comment (deyuan) [inline libpimeval/src/pimDevice.cpp:155]: We don't need to always show this if the env settings are optional

File: libpimeval/src/pimDevice.cpp
Comment (deyuan) [inline libpimeval/src/pimDevice.cpp:163]: Do we still support make PIM_SIM_TARGET?

**Review by deyuan — COMMENTED** [review]
Added some comments

**Comment by fasiddique** [comment]
I would like to test it before merging. Please let me know once it is ready to test and what are the steps. Until then, please hold off merging.

File: configs/PIMeval_BitSimdV.cfg
Comment (fasiddique) [reply configs/PIMeval_BitSimdV.cfg:7]: Does this mean that if hardware thread is 8 it will run using 8 cores and if the hardware thread is 32 it will use 10 cores?

Comment (fasiddique) [reply configs/PIMeval_BitSimdV.cfg:7]: Just tested the code and if the hardware has 12 hardware core it is using 10 cores. This is not correct behavior.

**Comment by fasiddique** [comment]
Thank you Hosein! This is really helpful addition to the simulator. I have tested the code and it works! However I have some concerns re behavior if a config file is not given, why are we showing this info and error? Is it intended? I think this is confusing and if no config file is given we should run it using the default and not show the warning.
```
PIM-Info: Trying to read it from envirnment variable PIMEVAL_TARGET
PIM-Info: Could not read environment variable PIMEVAL_TARGET
PIM-Error: Invalid simulation target 
```

File: configs/PIMeval_BitSimdV.cfg
Comment (mhgholamrezaei) [reply configs/PIMeval_BitSimdV.cfg:7]: @fasiddique Thanks for catching this issue. The right logic should be the following function:
num_threads = min(<user input>, std::thread::hardware_concurrency());
Please, correct me if I am wrong.

Comment (mhgholamrezaei) [reply configs/PIMeval_BitSimdV.cfg:7]: @deyuan, We can add "# ..." as line comment, however DRAMsim3 uses ";" for commenting. My suggestion is to support "#" for line comment and ";" as DRAMsim3 config files. What is you opinion?
[DRAMSim3 config comment example](https://github.com/umd-memsys/DRAMsim3/blob/29817593b3389f1337235d63cac515024ab8fd6e/configs/GDDR5_1Gb_x32.ini#L22C12-L22C50)

File: libpimeval/src/libpimeval.h
Comment (mhgholamrezaei) [reply libpimeval/src/libpimeval.h:33]: @deyuan Thank you for identifying this issue. Will be fixed.

File: libpimeval/src/pimSim.cpp
Comment (mhgholamrezaei) [reply libpimeval/src/pimSim.cpp:79]: @deyuan
Thanks for catching this, MLK. To resolve this issue, I propose adding a new private member field to the pimSim class to track whether init has been called. If init has already been called, an error status should be returned. Do you have any other suggestions?

File: configs/PIMeval_BitSimdV.cfg
Comment (fasiddique) [reply configs/PIMeval_BitSimdV.cfg:7]: > @fasiddique Thanks for catching this issue. The right logic should be the following function:
> 
> ```
> num_threads = min(<user input>, std::thread::hardware_concurrency());
> ```
> 
> Please, correct me if I am wrong.

Yes.
But what is the purpose of this parameter? Why aren't we directly taking the hardware thread count? Is this to sweep cores for simulator runtime?

File: libpimeval/src/pimDevice.cpp
Comment (mhgholamrezaei) [reply libpimeval/src/pimDevice.cpp]: @deyuan Thanks for noticing this bug. I solved it in new commit.

**Comment by mhgholamrezaei** [comment]
@deyuan @fasiddique 
Thanks for the code review.

In the last commit of this PR, the issues with simulation target PIM_DEVICE_UNKNOWN are solved. I need to discuss/clarify the following issues:
- How to resolve the value of thread_max?
- How to handle the MLK caused by `pimDevice::init()`?
- Comment syntax in the config files

**Comment by fasiddique** [comment]
Thanks Hosein for doing this! I pushed a minor change in the util to remove the warning and tested the code afterwards. Please review it. Everything else looks good to me! I am fine with merging this.

**Comment by mhgholamrezaei** [comment]
@fasiddique 
Thanks for fixing this issue.

**Review by fasiddique — APPROVED** [review]

File: libpimeval/src/pimDevice.cpp
Comment (deyuan) [inline libpimeval/src/pimDevice.cpp:286]: Where do you assign the m_deviceType variable?

File: libpimeval/src/pimSim.cpp
Comment (deyuan) [inline libpimeval/src/pimSim.cpp:126]: We need to move down this init call after the check below. If there is already a device, we should just return false without calling init or uninit.

File: libpimeval/src/pimSim.h
Comment (deyuan) [inline libpimeval/src/pimSim.h:154]: Can this be null by default, so that we use the default numbers in pimParamsDram ctor?

File: libpimeval/src/pimSim.cpp
Comment (deyuan) [inline libpimeval/src/pimSim.cpp:188]: We cannot call uninit here

**Review by deyuan — CHANGES_REQUESTED** [review]
Looks like there are a few issues need to be validated and resolved

File: libpimeval/src/pimSim.cpp
Comment (mhgholamrezaei) [reply libpimeval/src/pimSim.cpp:188]: Thanks. Updated.

File: libpimeval/src/pimSim.h
Comment (mhgholamrezaei) [reply libpimeval/src/pimSim.h:154]: Yes, It is updated.

File: libpimeval/src/pimDevice.cpp
Comment (mhgholamrezaei) [reply libpimeval/src/pimDevice.cpp:286]: It is still unclear for me how to resolve the value of m_deviceType based on m_simTarget. Could you please describe how these two variables are mapped together?

File: libpimeval/src/pimSim.cpp
Comment (mhgholamrezaei) [reply libpimeval/src/pimSim.cpp:126]: Thanks. Fixed!

**Comment by mhgholamrezaei** [comment]
@deyuan 
Currently, all the issues you mentioned have been resolved except for the one related to `m_deviceType`. Could you please review my previous message?

File: libpimeval/src/pimDevice.cpp
Comment (deyuan) [reply libpimeval/src/pimDevice.cpp:286]: Given a simTarget, we need to know if the application is running functional simulation or micro-ops level simulation. In functional simulation, the application can only use high-level functional APIs. In micro-ops level simulation, the application can only use architecture-specific micro-ops. We'll add device type sanity checks to limit the usage later. An application can create a device use PIM_FUNCTIONAL to run functional simulation and a specific device type to run micro-ops level simulation. For now, we can make sure the config file can only affect applications created with PIM_FUNCTIONAL for now.

**Comment by fasiddique** [comment]
@deyuan could you please check this out?

File: libpimeval/src/pimUtils.h
Comment (deyuan) [inline libpimeval/src/pimUtils.h:59]: It's better to keep just one unordered_map from enum to string. It's okay to be O(n) when converting from string to enum, since it is only used for config parsing.

File: libpimeval/src/pimUtils.h
Comment (deyuan) [inline libpimeval/src/pimUtils.h:74]: Is this being used?

File: libpimeval/src/pimSim.h
Comment (deyuan) [inline libpimeval/src/pimSim.h:154]: This variable is not needed

File: libpimeval/src/pimUtils.h
Comment (fasiddique) [reply libpimeval/src/pimUtils.h:59]: I actually think @mhgholamrezaei's approach is cleaner and we should keep it.

File: libpimeval/src/pimSim.cpp
Comment (deyuan) [inline libpimeval/src/pimSim.cpp:72]: Minor coding style: Suggest putting else inline for consistency and compactness, e.g. "} else {"

File: libpimeval/src/pimSim.h
Comment (deyuan) [inline libpimeval/src/pimSim.h:136]: Can we just keep one init function? The simConfig parameter can be empty string by default and act exactly like this

File: libpimeval/src/pimSim.cpp
Comment (deyuan) [inline libpimeval/src/pimSim.cpp:159]: Typo environment

File: libpimeval/src/pimSim.h
Comment (deyuan) [inline libpimeval/src/pimSim.h:137]: The parameter name is confusion and does not reflect it needs to be file contents. Can it just be file name string?

File: libpimeval/src/pimDevice.cpp
Comment (deyuan) [inline libpimeval/src/pimDevice.cpp:174]: Please assign value to m_deviceType, otherwise it's always NONE here

**Review by deyuan — COMMENTED** [review]
Thanks for the code changes. Request some further improvements with another round of review.

File: libpimeval/src/pimSim.h
Comment (deyuan) [reply libpimeval/src/pimSim.h:136]: You can keep this or change the other init file to take file name as param

File: libpimeval/src/pimUtils.h
Comment (deyuan) [reply libpimeval/src/pimUtils.h:59]: Okay, this is optional.

File: libpimeval/src/pimUtils.h
Comment (mhgholamrezaei) [reply libpimeval/src/pimUtils.h:74]: Yes. It is used in `pimDevice.cpp`.

File: libpimeval/src/pimSim.h
Comment (mhgholamrezaei) [reply libpimeval/src/pimSim.h:154]: Thanks. removed!

File: libpimeval/src/pimSim.cpp
Comment (mhgholamrezaei) [reply libpimeval/src/pimSim.cpp:72]: Thanks, Fixed these issues.

File: libpimeval/src/pimSim.cpp
Comment (mhgholamrezaei) [reply libpimeval/src/pimSim.cpp:159]: Thanks for catching this. Fixed.

File: libpimeval/src/pimSim.h
Comment (mhgholamrezaei) [reply libpimeval/src/pimSim.h:137]: Thanks. Renamed this variable to `simConfigFileConetnt`.

File: libpimeval/src/pimSim.h
Comment (mhgholamrezaei) [reply libpimeval/src/pimSim.h:136]: Fixed!

**Comment by mhgholamrezaei** [comment]
@deyuan 
Thanks for the code review. Please take a look at the new changes.

**Review by deyuan — APPROVED** [review]
Looks great to me, thanks!

**Pull request merged**
